cmd/hub/mcscontroller/controller.go

```diff
@@ -0,0 +1,169 @@
+		if apierrors.IsNotFound(err) {
+			log.Info("No ServiceExport found")
+		} else {
+			log.Error(err, "Error fetching ServiceExport")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, &se)
+	if err != nil {
+		return result, err
+	}
+
+	if se.GetDeletionTimestamp() != nil { // service export is marked to be deleted
+		return r.handleDelete(ctx, &se)
+	}
+
+	return result, err
+}
+
+// reconcile the cluster state.
+//
+// Reconciliation logic:
+// 1. list all relevant exports (i.e., having the same name and namespace labels)
+// 2. determine the primary export object
+// 3. iterate on all relevant export objects
+// 3a. if the object is being deleted (non empty deletion timestamp) - ignore
+// 3b. if object conflicts with primary - update export.Status.Condition.Conflict
+//		and delete corresponding ServiceImport.
+// 3c. otherwise, create/update the corrresponding ServiceImport and set the
+//		ServiceExport as owner.
+//
+// Notes:
+// - ensure objects are protected with a finalizer. Since we only receive the
+//	 affected object name into the reconciliation loop, we have no way to determine
+//	 and retrieve affected ServiceExports (i.e., client.List would not return the
+//   deleted object and we can't access its labels). The finalizer is cleared when we
+// 	 receive a notification for change in an object scheduled for deletion.
+// - ServiceImports are owned by the corresponding ServiceExport and are thus
+//	 automatically deleted by k8s when their ServiceExport is deleted.
+//
+func (r *ServiceExportReconciler) reconcile(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
+	lh := mcs.Lighthouse{}
+	name := lh.GetOriginalObjectName(se.ObjectMeta)
+	// @todo: add the NamespacedName of the ServiceExport to be compatible with Reconcile?
+	log := r.Log.WithValues("service", name).WithValues("cluster", lh.GetOriginalObjectCluster(se.ObjectMeta))
+
+	if !controllerutil.ContainsFinalizer(se, serviceExportFinalizer) && se.GetDeletionTimestamp() == nil {
+		controllerutil.AddFinalizer(se, serviceExportFinalizer)
+		if err := r.Client.Update(ctx, se); err != nil {
+			if apierrors.IsConflict(err) { // The SE has been updated since we read it, requeue to retry reconciliation.
+				return ctrl.Result{Requeue: true}, nil
+			}
+			log.Error(err, "Error adding finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+
+	//------------------------------------------
+	// -- @TODO CODE ABOVE SPECIFIED BEHAVIOR --
+	//------------------------------------------
+
+	return ctrl.Result{}, nil
+}
+
+// allow the object to be deleted by removing the finalizer.
+// Once all finalizers have been removed, the ServiceExport object and its
+// owned ServiceImport object will be deleted.
+//
+// Note: ownership will need to be reconsidered if merging imports
+func (r *ServiceExportReconciler) handleDelete(ctx context.Context, se *mcsv1a1.ServiceExport) (ctrl.Result, error) {
```

Comment:
why do we even need to bother with explicitly removing the finalizer? I recall you said that k8s removes objects with finalizers after some grace period.

Reply:
I was wrong. k8s will not delete the object until finalizers is empty